crates/sui-bridge-indexer/src/eth_worker.rs

```diff
@@ -57,204 +65,233 @@ impl EthBridgeWorker {
         })
     }
 
-    pub async fn start_indexing_finalized_events(
+    pub async fn sync_events(
         &self,
-        eth_client: Arc<EthClient<ethers::providers::Http>>,
-    ) -> Result<JoinHandle<()>> {
-        let newest_finalized_block = match get_latest_eth_token_transfer(&self.pg_pool, true)? {
-            Some(transfer) => transfer.block_height as u64,
-            None => self.config.start_block,
-        };
+        client: Arc<EthClient<ethers::providers::Http>>,
+        genesis_block: u64,
+    ) {
+        println!("Syncing event history");
 
-        info!("Starting from finalized block: {}", newest_finalized_block);
+        // get current block number
+        let current_block = self.provider.get_block_number().await.unwrap().as_u64();
 
-        let finalized_contract_addresses =
-            HashMap::from_iter(vec![(self.bridge_address, newest_finalized_block)]);
+        // get eth sync progress store
+        let (mut earliest_block_synced, latest_block_synced) =
+            read_eth_progress_store(&self.pg_pool).unwrap();
 
-        let (_task_handles, eth_events_rx, _) =
-            EthSyncer::new(eth_client, finalized_contract_addresses)
-                .run(self.bridge_metrics.clone())
-                .await
-                .map_err(|e| anyhow::anyhow!(format!("{:?}", e)))?;
+        if earliest_block_synced == 0 {
+            earliest_block_synced = current_block;
+        }
 
-        let provider_clone = self.provider.clone();
-        let pg_pool_clone = self.pg_pool.clone();
-        let metrics_clone = self.metrics.clone();
+        // start range from current block to MAX_BLOCK_RANGE blocks back
+        let mut end_block = current_block;
+        let mut start_block = end_block - MAX_BLOCK_RANGE;
 
-        Ok(spawn_logged_monitored_task!(
-            process_eth_events(
-                provider_clone,
-                pg_pool_clone,
-                metrics_clone,
-                eth_events_rx,
-                true
-            ),
-            "finalized indexer handler"
-        ))
-    }
+        loop {
```

Comment:
the logic in this loop is quite winding. It's also not clear to me how it enables parallel processing for multiple ranges at the same time.

Reply:
Agreed and it definitely doesn't enable parallel processing for multiple ranges... Open to how we can accomplish addressing both!

Reply:
Also because of rpc rate limiting, i'm not so sure that parallel processing for multiple ranges actually gets us better performance...